csrc/device_lower/utils.cpp

```diff
@@ -1013,6 +1016,775 @@ bool predicateAtEnd(ForLoop* loop) {
+// domain. From the projection, we will know if linear_g is linear and the
+// stride. For example, let's consider the following two schedules (same example
+// as the NVFuserTest.ProveLinearAndGetStride test):
+//
+// v1:
+//        I0         I1
+//       /  \       /  \.
+//          128        128
+//          / \        / \.
+//         /   \      /   \.
+//        /     \    /     \.
+//       /       \  /       \.
+//      /         \/        64.
+//     /          /\       /  \.
+//    /          /  \     /    \.
+//   16        [2]   8   8      8
+//                    \ /
+//                    xor
+//                    / \.
+//                   8   8
+//
+// v3:
+//        I0         I1
+//       /  \       /  \.
+//          32         256
+//          / \        / \.
+//         /   \      /   \.
+//        /     \    /     \.
+//       /       \  /       \.
+//      /         \/        64.
+//     /          /\       /  \.
+//    /          /  \     /    \.
+//   4          4    8   8      8
+//                    \ /
+//                    xor
+//                   /   \.
+//                  8     8
+//
+// Suppose that the [2] in v1 is linear_g, and the leaves in v3 are domain.
+// Domain is in the order [I0o, I1o, 4, 4, 8, 8, 8]. To figure out if linear_g
+// is linear in domain, we start from propagating linear_g back in v1. The first
+// step we see is the split of 128 into [2] and 64. From this split, we know
+// that when [2] projects to the 128, it projects to the 2 after a 64 on its
+// inner. The next step is the split of I1 by 128 in v1. Similarly, we know that
+// when [2] projects to I1, it projects to the 2 after a 64 on its inner.
+// Because the I1 in v1 and v3 are mapped, then we will continue propagation in
+// v3. The next step is to process the split of I1 into I1o and 256 in v3. We
+// already know that when [2] projects to I1, it projects to the 2 after a 64 on
+// its inner. Because 2 * 64 = 128, which is a factor of 256, we know that the
+// 256 is able to fully cover the [2] in I1. Therefore, I1o is unrelated to [2],
+// and we only need to focus on the 256. And when the [2] projects to this 256,
+// it projects to the 2 after a 64 on its inner. The next step is to process the
+// split of 256 into 4 and 64. Because the 64 happens to be the extent of the
+// inner of [2] in the 256, we know that the [2] will be fully covered by the
+// inner 2 of the 4 of the output of the split. So, we know that, when the [2]
+// projects to the 4, it projects to the inner 2 of the 4. Now we have finished
+// propagation and reached domain. Because [2] is the inner of the 4, and the 4
+// is linear in domain, so we have proved that [2] is linear in domain. In
+// domain, the domains on the right of the 4 are 8, 8, 8, so the stride is 8*8*8
+// = 512.
+namespace {
+
+// From the above example, we can see that how linear_g lives in domain could be
```

Comment:
General question: how is reordering handled? Since there's no "reorder" op, the ValGraph traversal itself doesn't seem to account for effects by reordering.

Reply:
Reorder does not need to be handled. Just like we do not need to handle reorder in the vectorization helper. However, we do check the orders when we see a merge, and we we run `proveLinearAndGetStrideAfterPropagation`(basically, any function that uses `search` checks the order).